### PR TITLE
Py open sci review

### PR DESCRIPTION
**Pull request by nilskre**
Fixed issues:
- #251 
- #252 
- #246 
- #195 
- #248 
- #249 
- #250 
- #258 
- #257 

Started working on #259 
Not yet finished and tested, because the trusted publisher mechanism can not be tested from a fork.
The rest of the PyOpenSci issues will be addressed within the next pull request.

### CODE REVIEW

File: .github/workflows/release.yaml [hunk 1]
```diff
@@ -1,17 +1,17 @@
-name: Docs build
-on:
--   push
+name: Release
+
+on: [push]
 
 jobs:
-    build:
+    build_and_deploy_docs:
```
Comment (slobentanzer) [inline .github/workflows/release.yaml:6]: @nilskre didn't we want to build the docs always, not just on a new release?

**Comment by slobentanzer** [comment]
@nilskre why do the macos runners cancel?

File: .github/workflows/release.yaml (hunk 1 above)
Comment (slobentanzer) [reply .github/workflows/release.yaml:6]: maybe a separate yaml for that would be easier to understand / maintain.

Comment (nilskre) [reply .github/workflows/release.yaml:6]: Yes, we can also build the docs always. I think for the deployment of the docs/artifact there are still some changes to be done within the next PR. So we could have a chat about this and then include the changes in the next PR. Would this be alright?

Comment (nilskre) [reply .github/workflows/release.yaml:6]: > maybe a separate yaml for that would be easier to understand / maintain.

Yes, I totally agree. So then three yaml files:
- ci_cd.yaml: test and code quality
- docs.yaml: build (and publish?) docs
- release.yaml: build and publish artifact

Are the names alright like this or could we improve understanding by other names?

Comment (slobentanzer) [reply .github/workflows/release.yaml:6]: > include the changes in the next PR. 

Sure, as long as all checks pass in this PR. I am now thinking that it would probably be easier to check the deploy functionality if we directed this PR at a feature branch (like `pyopensci`) instead of `main` directly. I have redirected the PR; we can merge it with lower risk.

Comment (slobentanzer) [reply .github/workflows/release.yaml:6]: > Are the names alright

Those are fine, and yes, publishing should also happen in the docs workflow.

**Comment by slobentanzer** [comment]
> the CICD workflow cannot push the new code coverage badge to the coverage branch in the original repo

yes. to be safe, i made a test PR, and the current workflow yaml has no issues pushing to the coverage branch. you could try removing some parts of the workflow iteratively to figure out what causes the permission error. the settings should be fine, the bot is allowed to read and write to branches.

**Comment by slobentanzer** [comment]
Relatedly, I am not sure we should push to the coverage branch on PRs at all. We can still compute, but it is not necessary to update the coverage from all PR CI runs. This should only be done on the `main` branch (meaning when merging something into `main`). Updating this would prevent this step from running in the PR, and thus prevent failure. We still need to make sure it will run on `main`, of course.

**Comment by nilskre** [comment]
Alright, then within the original repo, there are no problems. 
I have also checked to create a PR on my forked repo. All checks are also successful ([PR](https://github.com/nilskre/biocypher/pull/1)).

So I would assume, that the problems arise from creating a PR from a forked repo. I have not yet totally understood the problem. One related discussion about this problem is [here](https://github.com/ad-m/github-push-action/issues/96). So maybe it is related to the "Workflow permissions" of the Actions settings?

**Comment by nilskre** [comment]
> Relatedly, I am not sure we should push to the coverage branch on PRs at all. We can still compute, but it is not necessary to update the coverage from all PR CI runs. This should only be done on the `main` branch (meaning when merging something into `main`). Updating this would prevent this step from running in the PR, and thus prevent failure. We still need to make sure it will run on `main`, of course.

Yes, this makes sense. So we need something like a hook (to update the coverage branch), which is run after merging something into main. I am sure this is possible through Github Actions. I will have a look at it.

**Comment by nilskre** [comment]
Regarding the test coverage config:

tl;dr advantages of using [codecov](https://about.codecov.io/) for managing test coverage:
- Test coverage Dashboard
- PR comments
- free for Open Source projects

Explanation:
Currently, the test coverage is evaluated and the coverage badge is updated, but for new PRs, the change of test coverage is not considered.

In the aimed development scenario of BioCypher, where lots of different persons are developing BioCypher it would be convenient to have specific quality standards enforced for new PRs:
- Done: code quality checks
- Done: running the test suite
- Todo: monitor (and enforce?) a certain level of test coverage for new PRs

This could be achieved by using [codecov](https://about.codecov.io/). 
The usage of codecov is also recommended by the [pyOpenSci Guide](https://www.pyopensci.org/python-package-guide/ci-and-testing/intro.html) ([current proposal](https://github.com/lwasser/python-package-guide/blob/tests/ci-tests-data/code-cov.md)).

What do you think? Would it be an option to use codecov to monitor the test coverage?

**Comment by slobentanzer** [comment]
> Would it be an option to use codecov to monitor the test coverage?

Thanks for the summary. I looked this up briefly when I added the badge, and I decided to go with the "low tech" solution because it has less dependencies. We can consider codecov at a later time (not for this PR), you are welcome to open a dedicated issue for it. :)

There we could have a discussion about what exactly would be gained from using codecov; I am pretty sure that "enforcing" a specific coverage could be done in the low tech way as well. In addition, I don't think enforcing coverage is necessarily the best way to ensure good quality code.

**Comment by nilskre** [comment]
Yes, always good to follow the KISS principle and keep it simple.
I have created an issue for this #272 , so we can discuss this later. (:

File: .github/workflows/release.yaml (hunk 1 above)
Comment (nilskre) [reply .github/workflows/release.yaml:6]: To clarify the build and deployment of the docs:
- build the docs -> on every commit

Regarding deployment, I could imagine three scenarios:
[current] 1) Deploy the docs on every tag on the main branch and have one version of the docs: stable (state of the latest tag)
2) Deploy the docs on every commit on the main branch and have two versions of the docs: stable (state of the latest tag) and latest (latest commit on the main branch).
3) Deploy the docs on every commit on the main branch and have multiple versions of the docs: vX.X.X (docs for every version/tag), stable (state of the latest tag), and latest (latest commit on the main branch).

For now, I think option 1 is totally sufficient. But I could imagine, that possibility 2 or 3 could maybe be something for a future improvement?

So for now continue with option 1 and create an issue to discuss options 2 and 3. Would this be alright for you?

File: .github/workflows/ci_cd.yaml [hunk 2]
```diff
@@ -1,116 +1,155 @@
...
-        python-version: '3.9'
-
+      uses: actions/checkout@v4
     - name: Install and configure Poetry
       uses: snok/install-poetry@v1
       with:
         version: 1.5.1
-        virtualenvs-create: false
-
+        virtualenvs-create: true
```
Comment (slobentanzer) [inline .github/workflows/ci_cd.yaml:26]: @nilskre why do we switch to using venvs?

**Review by slobentanzer — APPROVED** [review]

File: .github/workflows/release.yaml (hunk 1 above)
Comment (slobentanzer) [reply .github/workflows/release.yaml]: Yes, I would keep it simple for now.

Comment (nilskre) [reply .github/workflows/release.yaml]: Great, then this is something to keep in mind when reaching the first major release of BioCypher.

File: .github/workflows/ci_cd.yaml (hunk 2 above)
Comment (slobentanzer) [reply .github/workflows/ci_cd.yaml:26]: Not particularly, just to keep it simple. These are good reasons, I just wanted to know why. :)

File: .github/workflows/ci_cd.yaml
Comment (slobentanzer) [inline .github/workflows/ci_cd.yaml:1]: @nilskre is there an option to use test.pypyi.org in the pyopensci branch to be able to test the release process first?

Comment (nilskre) [reply .github/workflows/ci_cd.yaml:1]: Yes, I have adapted the (former release.yaml) now publish.yaml to upload to test.pypi.org whenever a tag is set. After merging into the pyopensci branch we could set some tag there and test if the upload to test.pypi.org is successful. Is this a suitable procedure for testing?

Comment (slobentanzer) [reply .github/workflows/ci_cd.yaml:1]: yes, that was the basic idea. as soon as the CI passes, I will merge :)

Comment (nilskre) [reply .github/workflows/ci_cd.yaml:1]: Alright, perfect. Now everything is green (:
